daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala

```diff
@@ -72,322 +72,326 @@ abstract class SequenceParserBase(
-        newLastChildNode.get.isFinal = true
-        if (isOrdered) pstate.walker.walk()
-      }
-
-      scpIndex += 1
+        // We finished parsing one part of a sequence, which could either be an
+        // array, simple, or complex. We aren't sure if we actually added a new
+        // element or not, but in case we did, mark the last node as final.
+        //
+        // Additionally, if this is an ordered sequence, try to walk the infoset
+        // to output events for this potentially new element. If this is an
+        // unordered sequence, walking is unnecessary. This is because we may
+        // need to reorder the infoset once this unordered sequence is complete
+        // (via flattenAndValidateChildNodes below) and cannot walk until that
+        // happens. To ensure we don't walk even if a child parser tries to call
+        // walk() we incremented infosetWalkerBlockCount at the beginning of this
+        // function, so the walker is effectively blocked from making any
+        // progress. So we don't even bother calling walk() in this case.
+        if (newLastChildNode.isDefined) {
+          newLastChildNode.get.isFinal = true
+          if (isOrdered) pstate.walker.walk()
+        }
 
-    } // end while for each sequence child parser
+        scpIndex += 1
 
-    if (!isOrdered) {
-      // we are unordered, so we need to reorder the new children into schema
-      // definition order, flatten arrays, and validate
-      val infoset = pstate.infoset.asInstanceOf[DIComplex]
-      infoset.flattenAndValidateChildNodes(pstate, infosetIndexStart)
+      } // end while for each sequence child parser
 
-      // now that we have flattened children, we can decrement the block count
-      // that we incremented above. This will allow the infoset walker to walk
-      // into the new children that are now in the correct order.
-      pstate.infoset.infosetWalkerBlockCount -= 1
+      if (!isOrdered) {
+        // we are unordered, so we need to reorder the new children into schema
+        // definition order, flatten arrays, and validate
+        val infoset = pstate.infoset.asInstanceOf[DIComplex]
+        infoset.flattenAndValidateChildNodes(pstate, infosetIndexStart)
+
+        // now that we have flattened children, we can decrement the block count
+        // that we incremented above. This will allow the infoset walker to walk
+        // into the new children that are now in the correct order.
+        pstate.infoset.infosetWalkerBlockCount -= 1
+
+        // we've unblocked the unordered sequence, try walking to output
+        // everything we've created
+        pstate.walker.walk()
+      }
 
-      // we've unblocked the unordered sequence, try walking to output
-      // everything we've created
-      pstate.walker.walk()
+      if (child ne null) child.finalChecks(pstate, resultOfTry, priorResultOfTry)
+      ()
+    } finally {
+      pstate.mpstate.groupIndexStack.pop()
```

Comment:
This is mostly just indentation changes. If you click the gear at the top and click "Hide whitespace changes" you can see that the only thing that really changes is this new finally block.